File: Deploy/RapierLoom.Examples/16-Visitor/VB/Elements.vb
Public Class Expression
    Public Sub Accept(ByVal visitor As Visitor)
        visitor.Visit(Me)
    End Sub
End Class

Public Class Literal
    Inherits Expression

    Public Sub New(ByVal Value As Integer)
        _value = Value
    End Sub

    Public ReadOnly Property Value() As Integer
        Get
            Return _value
        End Get
    End Property

    Private _value As Integer
End Class

Public MustInherit Class Operation
    Inherits Expression

    Protected Sub New()
    End Sub

    Public MustOverride ReadOnly Property Symbol() As String
End Class

Public MustInherit Class BinaryOperation
    Inherits Operation

    Protected Sub New(ByVal left As Expression, ByVal right As Expression)
        Me._left = left
        Me._right = right
    End Sub

    Public ReadOnly Property Left() As Expression
        Get
            Return Me._left
        End Get
    End Property

    Public ReadOnly Property Right() As Expression
        Get
            Return Me._right
        End Get
    End Property

    Private _left As Expression
    Private _right As Expression
End Class


Public MustInherit Class UnaryOperation
    Inherits Operation

    Protected Sub New(ByVal expr As Expression)
        Me._expr = expr
    End Sub


    Public ReadOnly Property Expr() As Expression
        Get
            Return Me._expr
        End Get
    End Property

    Private _expr As Expression
End Class


Public Class Add
    Inherits BinaryOperation

    Public Sub New(ByVal left As Expression, ByVal right As Expression)
        MyBase.New(left, right)
    End Sub

    Public Overrides ReadOnly Property Symbol() As String
        Get
            Return "+"
        End Get
    End Property

End Class

Public Class Subt
    Inherits BinaryOperation
    Public Sub New(ByVal left As Expression, ByVal right As Expression)
        MyBase.New(left, right)
    End Sub

    Public Overrides ReadOnly Property Symbol() As String
        Get
            Return "-"
        End Get
    End Property

End Class

Public Class Neg
    Inherits UnaryOperation

    Public Sub New(ByVal exp As Expression)
        MyBase.New(exp)
    End Sub

    Public Overrides ReadOnly Property Symbol() As String
        Get
            Return "-"
        End Get
    End Property

End Class

File: Deploy/GripperLoom.Examples/16-Visitor/CS/Elements.cs
// ----------------------------------------------------------------
// * The LOOM.NET project http://rapier-loom.net *
// (C) Copyright by Wolfgang Schult. All rights reserved. 
// This code is licensed under the Apache License 2.0
// To get more information, go to http://loom.codeplex.com/license
// ----------------------------------------------------------------
using System;
using System.Collections.Generic;
using System.Text;

namespace VisitorExample
{
    /// <summary>
    /// The Expression accepts different visitors; Print and Eval
    /// </summary>
    public class Expression
    {
        public void Accept(Visitor visitor)
        {
            // here the visitors Visit() Method is called
            visitor.Visit(this);
        }
    }

    public class Literal : Expression
    {
        private int value;

        public int Value
        {
            get { return this.value; }
        }

        public Literal(int value)
        {
            this.value = value;
        }
    }

    public abstract class Operation : Expression
    {
        public abstract string Symbol { get; }
    }

    public abstract class BinaryOperation : Operation
    {
        protected BinaryOperation(Expression left, Expression right)
        {
            this.left = left;
            this.right = right;
        }

        private Expression left;
        public Expression Left
        {
            get { return left; }
        }

        private Expression right;
        public Expression Right
        {
            get { return right; }
        }

    }

    public abstract class UnaryOperation : Operation
    {
        Expression expr;

        public Expression Expr
        {
            get { return expr; }
        }

        protected UnaryOperation(Expression expr)
        {
            this.expr = expr;
        }
    }

    public class Add : BinaryOperation
    {
        public Add(Expression left, Expression right)
            :
            base(left, right)
        {
        }

        public override string Symbol
        {
            get { return "+"; }
        }
    }

    public class Sub : BinaryOperation
    {
        public Sub(Expression left, Expression right)
            :
            base(left, right)
        {
        }

        public override string Symbol
        {
            get { return "-"; }
        }

    }

    public class Neg : UnaryOperation
    {
        public Neg(Expression exp)
            : 
            base(exp)
        {
        }

        public override string Symbol
        {
            get { return "-"; }
        }

    }
}
